sdk/core/Azure.Core/src/Shared/BearerTokenChallengeAuthenticationPolicy.cs

```diff
@@ -0,0 +1,427 @@
+
+                if (getTokenFromCredential)
+                {
+                    if (backgroundUpdateTcs != null)
+                    {
+                        if (async)
+                        {
+                            info = await headerValueTcs.Task.ConfigureAwait(false);
+                        }
+                        else
+                        {
+#pragma warning disable AZC0104 // Use EnsureCompleted() directly on asynchronous method return value.
+                            info = headerValueTcs.Task.EnsureCompleted();
+#pragma warning restore AZC0104 // Use EnsureCompleted() directly on asynchronous method return value.
+                        }
+                        _ = Task.Run(() => GetHeaderValueFromCredentialInBackgroundAsync(backgroundUpdateTcs, info, context, async));
+                        return info.HeaderValue;
+                    }
+
+                    try
+                    {
+                        info = await GetHeaderValueFromCredentialAsync(context, async, message.CancellationToken).ConfigureAwait(false);
+                        headerValueTcs.SetResult(info);
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        headerValueTcs.SetCanceled();
+                        throw;
+                    }
+                    catch (Exception exception)
+                    {
+                        headerValueTcs.SetException(exception);
+                        throw;
+                    }
+                }
+
+                var headerValueTask = headerValueTcs.Task;
+                if (!headerValueTask.IsCompleted)
+                {
+                    if (async)
+                    {
+                        await headerValueTask.AwaitWithCancellation(message.CancellationToken);
+                    }
+                    else
+                    {
+                        try
+                        {
+                            headerValueTask.Wait(message.CancellationToken);
+                        }
+                        catch (AggregateException) { } // ignore exception here to rethrow it with EnsureCompleted
+                    }
+                }
+                if (async)
+                {
+                    info = await headerValueTcs.Task.ConfigureAwait(false);
+                }
+                else
+                {
+#pragma warning disable AZC0104 // Use EnsureCompleted() directly on asynchronous method return value.
+                    info = headerValueTcs.Task.EnsureCompleted();
+#pragma warning restore AZC0104 // Use EnsureCompleted() directly on asynchronous method return value.
+                }
+
+                return info.HeaderValue;
+            }
+
+            private (TaskCompletionSource<HeaderValueInfo> tcs, TaskCompletionSource<HeaderValueInfo>? backgroundUpdateTcs, bool getTokenFromCredential) GetTaskCompletionSources(TokenRequestContext context)
```

Comment:
IIRC, if you don't actually refer to the tuple names, don't use them. It pulls in an extra assembly for named value tuples on older frameworks, though that might predate netstandard2.0 (this caused a perf regression in an older codebase I worked on, which is how I found out about it).

It seems you just deconstruct the tuple anyway, so probably just best to avoid naming the members.

Reply:
I see a few other places in code you might want to avoid since you're deconstructing anyway.